### PR TITLE
Fix few bugs and add few improvements

### PR DESCRIPTION
**Pull request by silentlight**
Ticket: https://www.notion.so/figmentnetworks/Fix-problems-with-polkadot-proxy-b6027ee9b1cf4b92821c6804c73cabaa

- Fix bugs around sessions and eras
- Remove unneeded grpc arguments

### CODE REVIEW

File: grpc/block/blockpb/block.proto [hunk 1]
```diff
@@ -29,6 +28,7 @@ message Extrinsic {
 }
 
 message Block {
+  string block_hash = 1;
```
Comment (lukanus) [inline grpc/block/blockpb/block.proto:31]: May be nitpicking but shouldn't header be the very first field ? :)

**Review by lukanus — APPROVED** [review]

File: handlers/chain_handlers.js [hunk 2]
```diff
@@ -60,69 +62,45 @@ const getStatus = async (api, _call, context) => {
...
+  injectMetadata(api, currHeightMetadata);
 
-  let {blockHash, chain, specVersion} = data;
+  const {blockHash: currBlockHash} = currHeightMetadata;
 
-  let era = await api.query.staking.currentEra.at(blockHash);
-  let session = await api.query.session.currentIndex.at(blockHash);
+  const rawNextEra = await api.query.staking.currentEra.at(currBlockHash);
+  const nextEra = parseInt(rawNextEra.toString(), 10);
+  const rawNextSession = await api.query.session.currentIndex.at(currBlockHash);
```
Comment (pacmessica) [inline handlers/chain_handlers.js:92]: this is indbending. current is previous, and next is current? 🤯

File: handlers/chain_handlers.js [hunk 3]
```diff
@@ -60,69 +62,45 @@ const getStatus = async (api, _call, context) => {
...
-    era: prevEra,
-    session: prevSession,
-    lastInEra: prevEra !== currentEra,
-    lastInSession: prevSession !== currentSession,
-    chain: chain.toString(),
-    specVersion: specVersion.toString(),
-  };
-};
+  // Current height
+  // Current height block hash gets next era and session
```
Comment (pacmessica) [inline handlers/chain_handlers.js:83]: is it similar idea to this logic? https://github.com/paritytech/substrate-api-sidecar/blob/master/src/services/blocks/BlocksService.ts#L302

ie 
something like the block that's right before a new era/session falsely proclaims it's part of that new era/session?

File: handlers/height_handlers.js [hunk 4]
```diff
@@ -16,21 +16,20 @@ const getAll = async (api, call, context) => {
...
-  const blockResp = await blockHandlers.getByHeight(api, call, context);
-  const eventResp = await eventHandlers.getByHeight(api, call, context);
+  const [chainResp, blockResp, eventResp] = await Promise.all([
+    chainHandlers.getMetaByHeight(api, call, context),
+    blockHandlers.getByHeight(api, call, context),
+    eventHandlers.getByHeight(api, call, context),
+  ]);
 
   let stakingResp;
+  if (chainResp.lastInEra) {
```
Comment (pacmessica) [inline handlers/height_handlers.js:26]: 👍

File: utils/setup.js [hunk 5]
```diff
@@ -29,23 +31,34 @@ const fetchMetadataAtHeight = async (api, height) => {
...
+  return {
+    height,
+    blockHash,
+    ...metadataCache.items[runtimeVersionAt.specVersion]
+  };
+}
+
+// Return
+const getBlockHash = async (api, height) => {
+  if (height) {
```
Comment (pacmessica) [inline utils/setup.js:51]: what if height is 0?

File: utils/setup.js [hunk 6]
```diff
@@ -29,23 +31,34 @@ const fetchMetadataAtHeight = async (api, height) => {
     metadataCache.items[runtimeVersionAt.specVersion] = {
       chain,
       specVersion: runtimeVersionAt.specVersion,
-      height,
-      blockHash,
+      startedAtHeight: height,
+      startedAtBlockHash: blockHash,
```
Comment (pacmessica) [inline utils/setup.js:35]: is `startedAtBlockHash`  used anywhere?

and I'm trying to understand what `startedAt` prefix means. It's only `startedat` if we're indexing and calling heights one by one. If I call the proxy directly for data from a random height that the cache isn't aware of, then the cache will label it as `started` height, no?

File: grpc/block/blockpb/block.proto (hunk 1 above)
Comment (silentlight) [reply grpc/block/blockpb/block.proto:31]: I think here block_hash is good enough as first field ;)

File: handlers/chain_handlers.js (hunk 3 above)
Comment (silentlight) [reply handlers/chain_handlers.js:83]: `current` in this case means at the end of block and those values can be different at the beginning of block and at the end of block so it is better to take previous height in this case

File: handlers/chain_handlers.js (hunk 2 above)
Comment (silentlight) [reply handlers/chain_handlers.js:92]: yep, to get session and era you have to use previous block hash as explained above

File: utils/setup.js (hunk 6 above)
Comment (silentlight) [reply utils/setup.js:35]: `startedAtBlockHash` is not used anywhere. We can delete it. But in general we store the cache entry and we essentially store at what height the specVersion changes to a new value.

File: utils/setup.js (hunk 5 above)
Comment (silentlight) [reply utils/setup.js:51]: This is a valid concern. If height is 0 it should get finalizedHead. Good catch!

File: handlers/chain_handlers.js [hunk 7]
```diff
@@ -60,69 +62,45 @@ const getStatus = async (api, _call, context) => {
...
-  // Current height
-  injectMetadata(api, currHeightMetadata);
-
-  const {blockHash} = currHeightMetadata;
-  const rawCurrentEra = await api.query.staking.currentEra.at(blockHash);
-  const currentEra = parseInt(rawCurrentEra.toString(), 10);
-  const rawCurrentSession = await api.query.session.currentIndex.at(blockHash);
-  const currentSession = parseInt(rawCurrentSession.toString(), 10);
+const getMetaByHeight = async (api, call, context = {}) => {
+  const height = parseInt(call.request.height, 10);
```
[the remark this answers is not in the document]
Comment (silentlight) [reply handlers/chain_handlers.js:66]: nope, it is a string. It is weird since proto declares int64 :)

**Pull request merged**